demos/coreMQTT/mqtt_demo_keep_alive.c

```diff
@@ -0,0 +1,1177 @@
+        }
+
+        if( xDemoStatus == pdPASS )
+        {
+            /* Create an auto-reload timer to handle keep-alive. */
+            xKeepAliveTimer = xTimerCreateStatic( "KeepAliveTimer",
+                                                  mqttexampleKEEP_ALIVE_DELAY,
+                                                  pdTRUE,
+                                                  ( void * ) &xMQTTContext.transportInterface,
+                                                  prvKeepAliveTimerCallback,
+                                                  &xKeepAliveTimerBuffer );
+
+            if( xKeepAliveTimer == NULL )
+            {
+                LogError( ( "xTimerCreateStatic() failed to create KeepAliveTimer." ) );
+                xDemoStatus = pdFAIL;
+            }
+        }
+
+        if( xDemoStatus == pdPASS )
+        {
+            /* Start the timer for keep-alive. */
+            xTimerStatus = xTimerStart( xKeepAliveTimer, 0 );
+
+            if( xTimerStatus != pdPASS )
+            {
+                LogError( ( "xTimerStart() failed to start KeepAliveTimer." ) );
+                xDemoStatus = pdFAIL;
+            }
+        }
+
+        /****************************** Subscribe. ********************************/
+
+        if( xDemoStatus == pdPASS )
+        {
+            /* If server rejected the subscription request, attempt to resubscribe to
+             * topic. Attempts are made according to the exponential backoff retry
+             * strategy implemented in retryUtils. */
+            xDemoStatus = prvMQTTSubscribeWithBackoffRetries( &xMQTTContext );
+        }
+
+        /*********************** Publish and Receive Loop. ************************/
+
+        if( xDemoStatus == pdPASS )
+        {
+            /* Publish messages with QOS0, send and process keep-alive messages. */
+            for( ulPublishCount = 0; ulPublishCount < ulMaxPublishCount; ulPublishCount++ )
+            {
+                LogInfo( ( "Publish to the MQTT topic %s.", mqttexampleTOPIC ) );
+                xDemoStatus = prvMQTTPublishToTopic( &xMQTTContext );
+
+                if( xDemoStatus == pdPASS )
+                {
+                    /* Process incoming publish echo, since application subscribed
+                     * to the same topic the broker will send the publish message
+                     * back to the application. */
+                    LogInfo( ( "Attempt to receive publish message from broker." ) );
+                    xMQTTStatus = MQTT_ReceiveLoop( &xMQTTContext,
+                                                    mqttexampleRECEIVE_LOOP_TIMEOUT_MS );
+
+                    if( xMQTTStatus != MQTTSuccess )
+                    {
+                        LogError( ( "MQTT_ReceiveLoop() failed with status %s.",
+                                    MQTT_Status_strerror( xMQTTStatus ) ) );
+                        xDemoStatus = pdFAIL;
```

Comment:
If one iteration of the loop fails, should the loop continue executing?
In the `mqtt_demo_mutual_auth` demo, I have put the `xDemoStatus == pdPASS` condition in the condition expression of the loop to break out of the loop whenever failure occurs. (Alternatively, you can put a `break` here)

Reply:
Also, would be good to make similar changes in the plaintext and serializer demos

Reply:
The demo does not continue if one iteration fails.
https://github.com/aws/amazon-freertos/pull/2548/files#diff-e1a2055bf6afd86af80b7f6e76e1a28cR662